src/derive.rs

```diff
@@ -0,0 +1,143 @@
+use std::{fs::File, time::Duration};
+
+use egg::{RecExpr, Rewrite, Runner};
+
+use crate::*;
+
+pub fn parse<L: SynthLanguage>(filename: &str) -> Vec<Equality<L>> {
+    let file = File::open(filename).unwrap_or_else(|_| panic!("Failed to open {}", filename));
+    let report: Report<L> = serde_json::from_reader(file).unwrap();
+    let mut eqs = vec![];
+    eqs.extend(report.prior_rws);
+    eqs.extend(report.new_rws);
+    eqs
+}
+
+fn is_saturating<L: SynthLanguage>(lhs: &Pattern<L>, rhs: &Pattern<L>) -> bool {
+    let mut egraph: EGraph<L, SynthAnalysis> = Default::default();
+    let l_id = egraph.add_expr(&L::instantiate(lhs));
+    let initial_size = egraph.number_of_classes();
+
+    let r_id = egraph.add_expr(&L::instantiate(rhs));
+
+    egraph.union(l_id, r_id);
+    egraph.rebuild();
+    let final_size = egraph.number_of_classes();
+
+    initial_size >= final_size
+}
+
+fn mk_runner<L: SynthLanguage>(
+    egraph: EGraph<L, SynthAnalysis>,
+    lhs: &RecExpr<L>,
+    rhs: &RecExpr<L>,
+) -> Runner<L, SynthAnalysis> {
+    Runner::default()
+        .with_egraph(egraph)
+        .with_expr(lhs)
+        .with_expr(rhs)
+        .with_scheduler(egg::SimpleScheduler)
+        .with_hook(|r| {
+            if r.egraph.find(r.roots[0]) == r.egraph.find(r.roots[1]) {
+                Err("Done".to_owned())
+            } else {
+                Ok(())
+            }
+        })
+}
+
+pub fn derive<L: SynthLanguage>(params: DeriveParams) {
+    let test_rules = parse::<L>(&params.in1);
+    let num_test = test_rules.len();
+    let baseline_rules = parse::<L>(&params.in2);
+
+    println!("Using {} to derive {}", params.in1, params.in2);
```

Comment:
don't we want to do both directions?

Reply:
I don't think so- I think you should run it both ways if you want both ways, but there are legitimate use-cases (like Halide) where you want to know how many X rules can be derived by Y but you don't care about the reverse